### PR TITLE
Add an option in Savers and Removers to define whether events are thr…

### PR DESCRIPTION
**Pull request by MarieMinasyan**
| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | N/A
| Added Behats                      | N/A
| Changelog updated                 | N/A
| Review and 2 GTM                  | WIP
| Micro Demo to the PO (Story only) | N/A
| Migration script                  | N/A
| Tech Doc                          | N/A

…own from saveAll / removeAll or save / remove methods

### CODE REVIEW

**Comment by BitOne** [comment]
I'm not sure about that feature. It can be very strange that not all save trigger the events and generate difficult to understand bugs.

Can you give more details about what you're trying to fix with this PR ?

If it is about performances, the fix should maybe done on the subscribers themselves.

**Comment by MarieMinasyan** [comment]
The events are triggered the same way as before.
I simply added a new option, that allows to find out whether the PRE_SAVE and POST_SAVE events have been launched for a single entity update or not.

My case is: I need to call a third-party API. In case of a mass edit / an import / a rule execution, I should call a specific API with the ids of all impacted entities.

In case of a single entity update (from the UI), I should call another API.

Right now, there is no way of knowing in which case the event is launched.

**Comment by BitOne** [comment]
Okay, thanks @MarieMinasyan, didn't understand. It makes sense. ;)

**Comment by damien-carcel** [comment]
Rather than adding an option, wouldn't it be more efficient to add a new event?

- When dispatching `StorageEvents::PRE/POST_REMOVE` put data in a `RemoveEvent`
- When dispatching `StorageEvents::PRE/POST_REMOVE_ALL` put data in a `BulkRemoveEvent` (or `RemoveAllEvent` or whatever…)

**Comment by MarieMinasyan** [comment]
@damien-carcel I'm not sure, because I don't want to change the current behaviour.
And in some cases, may be it doesn't matter whether the event was dispatched from the bulk method or not.

**Comment by damien-carcel** [comment]
I didn't see your PR was on 1.6. Indeed an added option is probably better.

File: src/Pim/Bundle/CatalogBundle/Doctrine/Common/Saver/GroupSaver.php [hunk 1]
```diff
@@ -100,15 +100,15 @@ public function save($group, array $options = [])
     {
         $this->validateGroup($group);
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Doctrine/Common/Saver/GroupSaver.php]: `$options['unitary'] = true;` here to be consistent

File: src/Pim/Bundle/CatalogBundle/Doctrine/Common/Saver/GroupSaver.php [hunk 2]
```diff
@@ -120,10 +120,10 @@ public function saveAll(array $groups, array $options = [])
             return;
         }
 
-        $this->eventDispatcher->dispatch(StorageEvents::PRE_SAVE_ALL, new GenericEvent($groups));
-
         $options = $this->optionsResolver->resolveSaveAllOptions($options);
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Doctrine/Common/Saver/GroupSaver.php]: $options['unitary'] = false; here to be consistent

**Comment by rybus** [comment]
GTM

**Comment by jjanvier** [comment]
Sorry but I don't understand that 
> My case is: I need to call a third-party API. In case of a mass edit / an import / a rule execution, I should call a specific API with the ids of all impacted entities.

In that case you use the PRE_SAVE_ALL event. In the other case, you can call just PRE_SAVE. I don't understand why you need to add that option..

**Pull request merged**